File: server/src/core/extensions.py
from typing import Callable, Dict, List, Any


class Extensions:
    def __init__(self) -> None:
        self.extensions: List[Any] = []

    def use_extension(self, ext: Any) -> None:
        self.extensions.append(ext)

    def trigger_hook(self, name: str, context: Dict[str, Any] = {}) -> None:
        for ext in self.extensions:
            hook = getattr(ext, name, None)
            if callable(hook):
                hook(dict(context))  # pass a shallow copy

File: apps/nodejs/core/server/src/core/extensions.ts
import { ServerExtension } from "../types";



export class Extensions<TInjected extends object> {
    private extensions: ServerExtension<any>[] = [];

    useExtension<TNew extends object>(ext: ServerExtension<TNew>) {
        this.extensions.push(ext);
    }

    triggerHook(name: string, context: any = {}) {
        for (const ext of this.extensions) {
            const hook = ext[name as keyof typeof ext];
            if (typeof hook === 'function') {
                hook({ ...context });
            }
        }
    }
}
